promql/engine.go

```diff
@@ -1013,6 +1021,7 @@ func (ev *evaluator) eval(expr Expr) Value {
 
 	switch e := expr.(type) {
 	case *AggregateExpr:
+		unwrapParenExpr(&e.Param)
```

Comment:
I'd prefer not to mutate the main (non-cache etc.) fields of the AST during the query. E.g. a user might want to serialize or otherwise use it again after executing.

Could we make this functional?

```suggestion
		param := unwrapParenExpr(&e.Param)
```

Reply:
I agree that functional execution would be nicer.

However, evaluation already mutates the AST in other places, so actually making this functional would be a larger effort.

https://github.com/prometheus/prometheus/blob/master/promql/engine.go#L1056

Reply:
Re-evaluation might still produce correct results, but I wouldn't rely on it.

Reply:
Damn. My worst fears are confirmed ;) I think in this case it'd be really easy to avoid a mutation though, and fewer mutations cause fewer confusions.

Reply:
I quickly skimmed the code and found a couple places where evaluation has side effects on (non-cache) AST contents.

All of them look relatively fixable.

If functional execution is desired, I'd prefer to fix all of them in a separate PR and leave this one as is.

Reply:
Ok, I'd say let's put that on hold for now then and keep things as is here.

Reply:
I've made an issue for this #6627